src/compiler/factory.ts

```diff
@@ -2292,15 +2292,26 @@ namespace ts {
     /**
      * Sets a custom text range to use when emitting source maps.
      */
-    export function setSourceMapRange<T extends Node>(node: T, range: TextRange | undefined) {
+    export function setSourceMapRange<T extends Node>(node: T, range: SourceMapRange | undefined) {
         getOrCreateEmitNode(node).sourceMapRange = range;
         return node;
     }
 
     /**
+     * Create an external source map source file reference
+     */
+    export function createSourceMapSource(fileName: string, text: string, skipTrivia?: (pos: number) => number) {

```

Comment:
Leaving the result untyped avoids a catch-22 caused by https://github.com/Microsoft/TypeScript/blob/v2.3.4/src/services/types.ts#L70-L72.

I added return type and `as any` which I don't really like but there is no good way I found around this. Including `services/types.ts` (e.g. when building `tsserer.js`) requires this literal have `getLineAndCharacterOfPosition()` but not include `services/types.ts` (e.g. when building `tsc.js`) requires it not have it. This is caused indirectly because `SourceMapSource` needs to extend `SourceFileLIke` (which I changed to be more clear by explicitly extending the interface).